File: src/2017/day25/turing_machine.py
class Turing_machine:

	def __init__(self, Q, T, b, sigma, delta, q0, F):

		# finite set of states (set)
		self.Q = Q

		# finite set of tape alphabet symbols (set)
		self.T = T

		# blank symbol b ∈ T
		self.b = b

		# set of input symbol (subset of T) b ∉ Σ
		self.sigma = sigma

		# transition function 
		# δ:(Q∖F)×T→Q×T×{L,R}
		self.delta = delta

		# initial start state q0 ∈ Q
		self.q0 = q0

		# set of final/accepting states F ⊆ Q
		self.F = F


		self.current_state = q0
		self.tape = [self.b]
		self.rw_head = 0

	def step(self):
		# Execution
		# decode 
		write, move_dir, next_state = self.delta[self.current_state][self.tape[self.rw_head]]
		# Overwrite 
		self.tape[self.rw_head] = write
		# Move  
		if (move_dir == 'r'):
			if(self.rw_head == (len(self.tape) -1)):
				self.tape.append(self.b)
			self.rw_head += 1
		else:
			if (self.rw_head ==0):
				self.tape.insert(0, self.b)
			else:
				self.rw_head -= 1
		# Change state
		self.current_state = next_state
		pass
	def show_tape(self):
		print(self.tape)
	def checksum(self):
		return sum(self.tape)

File: src/2017/day25/turing_machine.cc
#include "turing_machine.h"

#include <iostream>
#include <numeric>
namespace TM {

Turing_machine::Turing_machine(std::set<char> Q, std::set<int> T, int b,
                               std::set<int> sigma, transition_data tranistions,
                               char q0, std::set<char> F) {
  this->Q = Q;
  this->T = T;
  this->b = b;
  this->sigma = sigma;
  this->q0 = q0;
  this->F = F;
  this->current_state = q0;
  this->rw_head = 0;
  this->transitions = tranistions;
  this->tape.push_back(this->b);
};
void Turing_machine::step() {
  step_data current_step = transitions.at(current_state).at(tape.at(rw_head));

  // decode
  auto [write, move_dir, next_state] = current_step;

  // overwrite
  auto it = tape.begin() + rw_head;

  tape.at(rw_head) = write;
  // move
  if (move_dir == 'r') {
    if (rw_head == tape.size() -1) {
      tape.push_back(b);
    }
    rw_head++;
  } else {
    if (rw_head == 0) {
      tape.push_front(b);
    } else {
      rw_head--;
    }
  }
  // change state
  this->current_state = next_state;
}  // namespace TM
void Turing_machine::show_tape() {
  auto it = tape.begin();
  auto end = tape.end();

  std::cout << "[";
  if (it != end) {
    std::cout << *it;
    ++it;
  }
  for (; it != end; ++it) {
    std::cout << " " << *it;
  }
  std::cout << "]\n";
}
int Turing_machine::checksum() {
  return std::accumulate(tape.begin(), tape.end(), 0);
}
}  // namespace TM
